### ISSUE TITLE
Trivy analyzer support

### ISSUE THREAD

**Issue by fnxpt** [DependencyTrack/dependency-track#3251]
### Current Behavior

No support for trivy

### Proposed Behavior

Trivy is one of the most performant analysers with really good results.
Dependency track should be able to use it and collect its results

### Checklist

- [X] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/master/CONTRIBUTING.md#filing-issues)
- [X] I have checked the [existing issues](https://github.com/DependencyTrack/dependency-track/issues) for whether this enhancement was already requested

**Comment by fnxpt** [comment]
I’m currently working on this and should have it finalised on the next couple of weeks

**Comment by lustr4** [comment]
Great, I'm waiting for this useful functionality (support of trivy analyzer).

**Comment by fnxpt** [comment]
I saw that issue but didn't want to consider it the same because I'm not building a generic analyzer... the plan for this is to use trivy server and do a similar thing done on snyk

**Comment by nscuro** [comment]
No worries @fnxpt, that's totally fine. Just wanted to have these linked as the other issue explicitly mentions Trivy support :)

**Comment by fnxpt** [comment]
Hey everyone quick update in here.

**Comment by fnxpt** [comment]
I created a PR for backend too... there are still some things that are not working as they should.

- Scans on projects that represent docker images are not retrieving OS vulnerabilities
- Scans on web projects are not working... probably due to empty groups

I still need to update the documention and add tests, but I wanted to share this as soon as possible to receive some feedback since this is my first PR in here.

**Comment by nscuro** [comment]
Thanks @fnxpt, this work is much appreciated.

> Regarding documentation I added a few things, but to be honest I'm really but in writing documentation, if someone would like to help let me know

Without having looked at it in too much detail, it would be good to have a (high-level) description of how this is supposed to be set up. To most it is not obvious that Trivy has a server mode, or what needs to be done to make it work. If we'd have a  step-by-step guide on how to set it up, including commands to launch Trivy, that would be super helpful.

Is there any public documentation on the Trivy API you're using?

**Comment by fnxpt** [comment]
Ok, I will try to add that information... additionally we can put as an option on the docker-compose, what do you think @nscuro .
Regarding the API i couldn't find any documentation I had to check their code. They have 2 types of API. A REST API and a protobuf API.

**Comment by fnxpt** [comment]
All package managers that trivy supports should be supported now, docker is also supported now.
@nscuro - is there any way to generate a snapshot image, I would like to still use the official branch for that but if its not possible I will generate one on my fork.
Next steps: Test and compare the results against trivy, just to see if we are missing something... do a little bit of clean up and then add the missing tests.

**Comment by nscuro** [comment]
@fnxpt I pushed images for your changes here:

```
dependencytrack/apiserver:3259-trivy-integration
dependencytrack/frontend:656-trivy-integration
```

**Comment by fnxpt** [comment]
Thanks @nscuro 

So if someone wants to try it my advice is to use the images above + a container running trivy server.
After that you just need to enable it on the Admin -> Analysers settings by providing the endpoint + token.

**Comment by nscuro** [comment]
The following `docker-compose.yml` can be used for the test setup:

```yml
services:
  apiserver:
    image: dependencytrack/apiserver:3259-trivy-integration
    ports:
    - "127.0.0.1:8080:8080"
    volumes:
    - "apiserver-data:/data"
    restart: unless-stopped

  frontend:
    image: dependencytrack/frontend:656-trivy-integration
    environment:
      API_BASE_URL: "http://localhost:8080"
    ports:
    - "127.0.0.1:8081:8080"
    restart: unless-stopped

  trivy:
    image: aquasec/trivy:latest
    command:
    - server
    - --listen
    - 0.0.0.0:8080
    - --token
    - dummy
    restart: unless-stopped

volumes:
  apiserver-data: { }
```

Trivy should then be configured as follows (token is `dummy`):

<img width="826" alt="image" src="https://github.com/DependencyTrack/dependency-track/assets/5693141/aeb926fc-b782-4878-9f04-875e8b5062ae">

I would also recommend to disable OSS Index for the duration of the test, as OSS Index is invoked prior to Trivy and there may be significant overlap in findings. Disabling OSS Index makes it easier to see how many findings Trivy is able to raise.

<img width="826" alt="image" src="https://github.com/DependencyTrack/dependency-track/assets/5693141/f9640934-5720-4f61-89e4-aeb3fd811402">

**Comment by fnxpt** [comment]
Can you check the logs from both the api-server and trivy server.... the scan is fast (less than 10s), so problably its a connection issue

**Comment by simonfrancaix** [comment]
@fnxpt The logs don't show any connection problems between the api-server and the trivy server. By importing the bom again I managed to obtain vulnerabilities for my components. I import 2 different boms which seems to be the cause of my problem.

For this bom I do have some vulnerabilities that have come up:

`trivy image --format cyclonedx --output traefik.json traefik:2.10`

For this bom I have no vulnerabilities:

`trivy image --format cyclonedx --output docker-bom/quarkus.json registry.access.redhat.com/ubi8/openjdk-17:1.17-1.1696520325`

For the second I don't see any trivy analysis triggered by the api-server. The difference between these 2 boms is that the second only has system dependencies. When using Trivy, I expect Dependency Track to report a vulnerability in the `curl` package of the `registry.access.redhat.com/ubi8/openjdk-17:1.17-1.1696520325` image.

![image](https://github.com/DependencyTrack/dependency-track/assets/119447517/161a5f7b-9aa4-4d94-be75-3fa094bff33f)

**Comment by fnxpt** [comment]
@simonfrancaix if the connection is established it should have 3 log entries for the 3 endpoints called
it should be something like PutBlob 200, Scan 200, DeleteBlob 200

**Comment by fnxpt** [comment]
Theres a bug for the second one, Im working on it. The problem is related to the package manager declared. give me 1h and I should be able to submit a commit

**Comment by simonfrancaix** [comment]
Thanks for the quick response!

FYI by running the trivy analysis directly on the image I find this result:

```
registry.access.redhat.com/ubi8/openjdk-17:1.17-1.1696520325 (redhat 8.8)
=========================================================================
Total: 141 (UNKNOWN: 0, LOW: 82, MEDIUM: 56, HIGH: 3, CRITICAL: 0)
```

I'll test again at the end of the day. Don't hesitate to let me know as soon as the 3259-trivy-integration tag is updated.

**Comment by simonfrancaix** [comment]
If I inspect the result of trivy in CLI I find these vulnerabilities for `cups-libs`:

![image](https://github.com/DependencyTrack/dependency-track/assets/119447517/f3e44f05-7f8b-4475-8914-6a9efa7f8c26)

Is it possible that trivy's database is different in server mode than in CLI?

**Comment by fnxpt** [comment]
No it's something else, I'm trying to validate what's the problem

**Comment by fnxpt** [comment]
Sorry for the delay, there were 3 issues.
1- isCapable was returning false when purl was not filled or with a missing package type
2- Arch was hardcoded
3- Epoch was not filled (this was the main issue for the additional vulnerabilities)

**Comment by nscuro** [comment]
@fnxpt I pushed an updated image built against 62528bbfd85c2b80eb56f80569e9b77c10fd4b33 to Docker Hub.

**Comment by simonfrancaix** [comment]
Sorry for the delay, it's been a busy day.

The various modifications have corrected the problem, thank you again for your responsiveness.

For test I used a bom generated as follows:

`trivy image --format cyclonedx --output ubi.json registry.access.redhat.com/ubi8/openjdk-17:1.17-1.1696520325`

Dependency Track analysis results:

![image](https://github.com/DependencyTrack/dependency-track/assets/119447517/533c9c44-985d-4d7f-b56a-f12379b07ce2)

In Dependency Track we detect an additional vulnerability in Trivy CLI.

![image](https://github.com/DependencyTrack/dependency-track/assets/119447517/d992de69-1961-436d-8327-56b53bbc7205)

The additional vulnerability concerns this component:

![image](https://github.com/DependencyTrack/dependency-track/assets/119447517/5977a7e6-f83b-4039-81cf-6bc99b80dd0a)

Nor does it appear to count against the vulnerability counter in the top right-hand corner. This problem is certainly not related to your feature.

![image](https://github.com/DependencyTrack/dependency-track/assets/119447517/1a308a79-7b38-4002-959d-931e6764409a)

![image](https://github.com/DependencyTrack/dependency-track/assets/119447517/5d595dff-4f4c-4bdd-8b8e-eb091032fa4c)

I also carried out a second test. For some of our projects we try to provide a single bom to Dependency Track to provide an overview for a project. It's certainly not ideal, but I haven't found a better solution for the moment for aggregating the various vulnerabilities of a project.

To keep things simple, I simply merged two boms representing two containers to evaluate this feature. It seems that only the first container detected is considered by the analysis.

```bash
trivy image --format cyclonedx --output traefik.json traefik:2.10
trivy image --format cyclonedx --output ubi.json registry.access.redhat.com/ubi8/openjdk-17:1.17-1.1696520325
cyclonedx-cli merge --input-files ubi.json traefik.json --output-file bom.json --name my-app --version latest --hierarchical
```

![image](https://github.com/DependencyTrack/dependency-track/assets/119447517/2f0c91fd-3350-4440-8192-3e3059822d1d)

![image](https://github.com/DependencyTrack/dependency-track/assets/119447517/ff582f8e-15ca-46f4-b767-4e622d15cd61)

If I run a new scan using the "Reanalyze" button in the audit vulnerabilities menu, I find all the vulnerabilities (144 for ubi and 5 for traefik).

![image](https://github.com/DependencyTrack/dependency-track/assets/119447517/b87980f5-976e-42b1-aee2-3a2248ef6f3d)

The logs show that this time it was the other image that was analyzed.

![image](https://github.com/DependencyTrack/dependency-track/assets/119447517/421c9b08-d4c0-411f-87a6-248d6d8523be)

BTW Your development of the trivy analyzer is even better than trivy itself - impressive! Let me tell you why. If I run a scan on the shared bom with trivy, it doesn't find any vulnerabilities.

![image](https://github.com/DependencyTrack/dependency-track/assets/119447517/91ac5b3c-c8d2-4a4d-8b3f-6d7c70e6984f)

Sorry for the long comment, I hope my tests help you implement this great feature.

**Comment by fnxpt** [comment]
Thanks for the feedback regarding the additional vulnerability I will need to have a look.
Regading the better results, it might be because trivy cli doesnt expect merged sboms, since we are doing it on our end to we also included it... I just need to double check the final package because maybe we are combining the packages into the wrong OS.
I will check these 2 things, if you have anything extra to validate let me know

**Comment by romwol-witt** [comment]
Hi! First of all, seems to be great work 👍 But i cannot see at the logs that the TrivyAnalyzer is started... Any idea how to debug? I have a trivy-operator in Server-Mode running at my cluster and configured it in DT... 
![Bildschirmfoto 2023-12-11 um 10 22 10](https://github.com/DependencyTrack/dependency-track/assets/152502283/d5cec380-3fad-4956-918c-e7e5007e915d)

**Comment by fnxpt** [comment]
@romwol-witt you should be able to see a `Starting Trivy vulnerability analysis task` message in the logs. Also when the requests are made to the server you should see some PutBlob/Scan/DeleteBlob messages.
Best way to validate it is to force a scan in one of the projects or add a new SBOM to a project.

**Comment by romwol-witt** [comment]
Ok, many thanks. Now its working... But if i compare the results with the results from trivy operator, there is a HUGE gap regarding the amount of vulnerabilities :( Any idea why?

**Comment by fnxpt** [comment]
Can you share your findings… I’m trying to identify issues.Sent from my iPhoneOn 12 Dec 2023, at 08:26, romwol-witt ***@***.***> wrote:﻿
Ok, many thanks. Now its working... But if i compare the results with the results from trivy operator, there is a HUGE gap regarding the amount of vulnerabilities :( Any idea why?

**Comment by fnxpt** [comment]
Can you share your findings I'm trying to fix this type of issues

**Comment by romwol-witt** [comment]
Oh, i see there is a problem in the log... hm... 

`2023-12-12 08:20:29,498 ERROR [TrivyAnalysisTask] Request failure
org.apache.http.conn.ConnectionPoolTimeoutException: Timeout waiting for connection from pool`

--------------
There are findings for other projects, but there not as much as in trivy as well. Here are the screenshots of the difference:

Trivy Operator:
<img width="840" alt="Bildschirmfoto 2023-12-12 um 09 27 54" src="https://github.com/DependencyTrack/dependency-track/assets/152502283/0f632683-55f4-4a69-a337-41a587665d1e">

Dtrack:
![Bildschirmfoto 2023-12-12 um 09 28 24](https://github.com/DependencyTrack/dependency-track/assets/152502283/ac0ecedc-7aed-4e6f-bcee-f0e6af3f8de4)

**Comment by fnxpt** [comment]
That's weird, can you check the logs from trivy server. Can you share the sbom, this way I can validate on my side too.

**Comment by romwol-witt** [comment]
[node-sbom.json](https://github.com/DependencyTrack/dependency-track/files/13645873/node-sbom.json)

**Comment by romwol-witt** [comment]
now i have findings for node?!?! But why?! ^^ But still not the same as trivy operator. i cannot use the frontend at the moment because the apiserver doing a lot of stuff and is not responding to the frontend. i'll keep you updated if i can reach the frontend again

**Comment by fnxpt** [comment]
The sbom you sent is not retrieving any components to dependency track so its not able to be processed

**Comment by romwol-witt** [comment]
I have a script that extracts the component part and that i am uploading to DT... Like this:
[librarynode_bom.json](https://github.com/DependencyTrack/dependency-track/files/13646632/librarynode_bom.json)

**Comment by fnxpt** [comment]
I see results, for that sbom, but for sure something is missing I will need to debug it again
<img width="1389" alt="Screenshot 2023-12-12 at 11 04 06" src="https://github.com/DependencyTrack/dependency-track/assets/1815240/9020f1b1-77fd-43e8-9f4a-30c3ab94fd0d">
<img width="569" alt="Screenshot 2023-12-12 at 11 05 07" src="https://github.com/DependencyTrack/dependency-track/assets/1815240/a3c957ed-ee4f-4db7-978b-dc816b8d6840">

**Comment by romwol-witt** [comment]
I've tried https://github.com/ckotzbauer/sbom-operator as well and there i get much more results than from trivy ^^

![Bildschirmfoto 2023-12-12 um 11 40 48](https://github.com/DependencyTrack/dependency-track/assets/152502283/9828a90a-b968-4e24-bbd5-6b637353773d)

**Comment by fnxpt** [comment]
So the goal is to have the same results on dependency track as if you were running trivy tool directly, from the previous sbom you sent there is clearly some differences, I will need to understand why

**Comment by romwol-witt** [comment]
if i can help in any way, just let me know! I would prefer the trivy way also

**Comment by romwol-witt** [comment]
Btw, how do you get the sbom's to dependency track? i am using a bash script currently, but this is far away from optimal...

**Comment by simonfrancaix** [comment]
@romwol-witt I don't think this ticket is the right place to expand on automating sbom publishing in Dependency Track. Quite simply, I think automation (CI/CD) is the key. For example, we run a pipeline for each project, generating the SBOM and publishing it to Dependency Track. This pipeline runs automatically every day from the main development branch, providing our developers with the dependency status every day. We don't currently use Kubernetes to orchestrate our containers, but it's certainly simple enough to run a kubernetes job periodically to update your project's sbom. This job should retrieve the sbom generated by trivy-operator and then upload it to Dependency Track.

@fnxpt I'll continue to integrate the feature with more complex projects and also look at its integration with other scanners. Let me know if I can assist you in solving the sbom merge analysis.

**Comment by fnxpt** [comment]
@romwol-witt you can join slack channel and discuss this questions in there I will gladly give some hints

@simonfrancaix right now Im trying to identify the existing issues... one way to help is to provide sboms where the outcome is different from running trivy cli

**Comment by simonfrancaix** [comment]
@fnxpt here's a bom that returns a different result [ubi.json](https://github.com/DependencyTrack/dependency-track/files/13652064/ubi.json)

Component: pkg:rpm/redhat/guava@31.0.1-4.module%2Bel8.8.0%2B18044%2B0a924b8f?arch=noarch&distro=redhat-8.8&modularitylabel=maven%3A3.8%3A8080020230202141436%3A1e199b36

![image](https://github.com/DependencyTrack/dependency-track/assets/119447517/af6f3d69-60cf-4e04-a69d-b0796e15dbd7)
![image](https://github.com/DependencyTrack/dependency-track/assets/119447517/77ac4679-aeb0-4c40-879f-7ed774bc2bf9)
![image](https://github.com/DependencyTrack/dependency-track/assets/119447517/8d86da0f-3094-4f6e-8753-9426ec000e1c)

---

I left this version of dependency track running in order to visualize the appearance of new vulnerabilities for a project. Example:

![image](https://github.com/DependencyTrack/dependency-track/assets/119447517/e81aca8b-9088-4082-af79-255b1b6db869)

This new vulnerability is also present in the new analysis of trivy-cli

**Comment by romwol-witt** [comment]
@simonfrancaix Sorry, you're completely right. I just wanted to know how you guy get the trivy-sboms to dependency track. But i will join the slack channel for this...

**Comment by fnxpt** [comment]
@simonfrancaix executed it locally and got the same amount of vulnerabilities.. maybe its because of the latests changes I did (still not commited)
<img width="898" alt="Screenshot 2023-12-13 at 10 19 40" src="https://github.com/DependencyTrack/dependency-track/assets/1815240/4a6a935a-ccad-4a6b-bfcb-5e8fbd9bf8cd">
<img width="564" alt="Screenshot 2023-12-13 at 10 19 24" src="https://github.com/DependencyTrack/dependency-track/assets/1815240/9459a6be-3941-42fe-88bc-18f28d7b8525">

**Comment by fnxpt** [comment]
@romwol-witt figure out the issue with node sbom... 
For example imagemagic is retrieving vulnerabilities but imagemagick-6-common is not.
When calling trivy directly with sbom the file contains some properties describing the srcName as imagemagic, internally trivy scans imagemagik instead retrieving the same vulnerabilities for imagemagick-6-common.
While we dont have the same amount of vulnerabilities I think this is ok because the vulnerabilities will still be there and you wont be able to fix imagemagick-6-common without fixing imagemagick.
@nscuro what do you think?

**Comment by jimmyjones2** [comment]
Having Trivy will be great!

I've tried with the `docker-compose.yml` above, enabling the trivy analyzer and setting base URL to `http://trivy:8080` and API token `dummy`. The results don't seem quite right and I'm getting the following errors in the log when uploading an SBOM created by trivy (cyclonedx json):

```
2023-12-31 14:05:03,519 INFO [TrivyAnalysisTask] add package pkg:rpm/redhat/zlib@1.2.7-20.el7_9?arch=x86_64&distro=redhat-7.9
2023-12-31 14:05:03,559 INFO [TrivyAnalysisTask] PutBlob response: 200
2023-12-31 14:05:04,204 INFO [TrivyAnalysisTask] Scan response: 200
2023-12-31 14:05:04,251 INFO [TrivyAnalysisTask] received response from trivy
2023-12-31 14:05:04,251 WARN [TrivyParser] Unable to parse published date 2017-09-06T21:29:00.707Z
2023-12-31 14:05:04,251 WARN [TrivyParser] Unable to parse last modified date 2017-09-13T11:17:22.230Z
2023-12-31 14:05:04,305 INFO [TrivyAnalysisTask] Trivy vulnerability added : CVE-2015-5186 to component audit-libs
2023-12-31 14:05:04,310 WARN [TrivyParser] Unable to parse published date 2019-06-18T18:15:09.247Z
2023-12-31 14:05:04,310 WARN [TrivyParser] Unable to parse last modified date 2023-11-07T02:13:33.410Z
2023-12-31 14:05:04,389 INFO [TrivyAnalysisTask] Trivy vulnerability added : CVE-2012-6711 to component bash
```

**Comment by fnxpt** [comment]
The errors are related with the parsing of the dates, I notice that and I will fix it soon, regarding the results what do you think its wrong, did you compare it with trivy. If you provide the sbom file I can also have a look

**Comment by fnxpt** [comment]
Issue with date parsing fixed on last commit

**Comment by jimmyjones2** [comment]
![Screenshot from 2024-01-01 19-37-22](https://github.com/DependencyTrack/dependency-track/assets/2991805/c2890064-ca15-4cbc-a690-aba4c36f8bd9)
Just noticed the links to the trivy CVEs need to be lowercased - otherwise get a 404

**Comment by jimmyjones2** [comment]
Generated an SBOM for a Red Hat UBI8 image (latest as of today):
```
trivy image --format cyclonedx --output result.json registry.access.redhat.com/ubi8/ubi:8.9-1028
```
Standard scan, showing ~183 vulns, as Red Hat rate many as unaffected etc - essentially useless
```
trivy sbom result.json
```
Ignore unfixed, which results in 5 CVEs in 2 packages - actionable (wait until Red Hat produce new images, or possibly run `dnf update` as part of Dockerfile)
```
trivy sbom result.json --ignore-unfixed
```

If I create 3 new projects, and upload the `result.json` created above to each of them, get different results for each. Going into each shows 190 audit vulnerabilities, which isn't helpful. One of the benefits of Trivy is it includes the Red Hat feed, so you can filter (e.g. with `--ignore-unfixed`) to make the results actionable. Could the Trivy analyzer add support for this please?
![image](https://github.com/DependencyTrack/dependency-track/assets/2991805/3819f661-5c9a-4082-9a41-8ef805483c77)

**Comment by fnxpt** [comment]
```trivy image registry.access.redhat.com/ubi8/ubi:8.9-1028``` returned this
<img width="549" alt="Screenshot 2024-01-01 at 21 11 54" src="https://github.com/DependencyTrack/dependency-track/assets/1815240/6aa0f2cf-0e32-4611-a703-248976862d81">
generated and imported sbom into dependency track and I get the same vulnerabilities
<img width="879" alt="Screenshot 2024-01-01 at 21 18 06" src="https://github.com/DependencyTrack/dependency-track/assets/1815240/75ed6cca-d4d3-4b44-92d4-950f323c46e9">

**Comment by fnxpt** [comment]
> ![Screenshot from 2024-01-01 19-37-22](https://private-user-images.githubusercontent.com/2991805/293604233-c2890064-ca15-4cbc-a690-aba4c36f8bd9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQxNDAxMzEsIm5iZiI6MTcwNDEzOTgzMSwicGF0aCI6Ii8yOTkxODA1LzI5MzYwNDIzMy1jMjg5MDA2NC1jYTE1LTRjYmMtYTY5MC1hYmE0YzM2ZjhiZDkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEwMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMDFUMjAxMDMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Y2VhMzljMDBlYjhhOTI2ZWI0NWU0YjM3YzMyZDI1Y2I0MWRiNDk4NGVhYzVmOTFiY2NiYjE2MjRkMjliNDQxMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.74gMIogcxgaS7v_99hyluiZki2xwUtzbpuSsWA6rJG4) Just noticed the links to the trivy CVEs need to be lowercased - otherwise get a 404

The problem is not the case, but the link it self, some of the CVEs don't exist on aquasecurity side, but also the format of the link is wrong... Im going to create a PR to change it to nist page instead

**Comment by fnxpt** [comment]
I ended up implementing it..
<img width="1231" alt="Screenshot 2024-01-01 at 22 02 05" src="https://github.com/DependencyTrack/dependency-track/assets/1815240/e4e5fa46-0b7d-4bbb-b1dd-6b043e800ec6">

**Comment by jimmyjones2** [comment]
Thanks @fnxpt 

I presume this issue of multiple uploads of the same SBOM to different projects showing different number of vulnerabilities is an issue in DependencyTrack and not anything to do with the Trivy scanner?

![image](https://github.com/DependencyTrack/dependency-track/assets/2991805/3819f661-5c9a-4082-9a41-8ef805483c77)

**Comment by fnxpt** [comment]
Probably its not an issue at all, but a job that didn't run yet... the jobs are performed in x hours, so probably they were not updated yet

**Comment by jimmyjones2** [comment]
Sorry, yes I was being too impatient, they all now show the same. Thanks for your work on Trivy, this will make DependencyTrack usable for people using Red Hat container images.

**Comment by nscuro** [comment]
> @fnxpt For example imagemagic is retrieving vulnerabilities but imagemagick-6-common is not.
When calling trivy directly with sbom the file contains some properties describing the srcName as imagemagic, internally trivy scans imagemagik instead retrieving the same vulnerabilities for imagemagick-6-common.
While we dont have the same amount of vulnerabilities I think this is ok because the vulnerabilities will still be there and you wont be able to fix imagemagick-6-common without fixing imagemagick.
@nscuro what do you think?

Yes I think that is fine. If I got this right, it also means that scanning an SBOM with Trivy that was not *generated* by Trivy, this kind of metadata will also not be there? 

Can we forward arbitrary properties to the Trivy API, assuming we have them? I think once #2560 lands, we should be able to simply present the properties ingested from the BOM to Trivy.

**Comment by fnxpt** [comment]
> > @fnxpt For example imagemagic is retrieving vulnerabilities but imagemagick-6-common is not.
> > When calling trivy directly with sbom the file contains some properties describing the srcName as imagemagic, internally trivy scans imagemagik instead retrieving the same vulnerabilities for imagemagick-6-common.
> > While we dont have the same amount of vulnerabilities I think this is ok because the vulnerabilities will still be there and you wont be able to fix imagemagick-6-common without fixing imagemagick.
> > @nscuro what do you think?
> 
> Yes I think that is fine. If I got this right, it also means that scanning an SBOM with Trivy that was not _generated_ by Trivy, this kind of metadata will also not be there?
> 
> Can we forward arbitrary properties to the Trivy API, assuming we have them? I think once #2560 lands, we should be able to simply present the properties ingested from the BOM to Trivy.

Yes if we have the properties on DT we can use them when performing the requests, those properties are filled on the sbom but we don't have any way to access them. If in the future we do we can also import them, but to be honest since you will only need to fix the main library I think its a nice to have.
Maybe we can merge this one and create a issue tied to #2560

**Comment by nscuro** [comment]
> but to be honest since you will only need to fix the main library I think its a nice to have.
Maybe we can merge this one and create a issue tied to https://github.com/DependencyTrack/dependency-track/issues/2560

Agreed on the nice-to-have point. Yes, let's raise a follow-up issue for that.

**Comment by sithmein** [comment]
I was just trying this on a Docker image based on Ubuntu. There are several CVEs reported for `libc6` and `libc-bin` when running trivy directly. However, when using an SBOM via DT there are no vulnerabilities for libc being reported. However, I am also getting the date parsing errors mentioned above. Could this be the reason for the missing findings? I can provide more information if required.
Is there a rough timeline when this can get integrated? Especially when it comes to OS support trivy will be very useful as OSSIndex doesn't have much data about it.

**Comment by nscuro** [comment]
@sithmein @fnxpt has made quite a few adjustments and fixes since I last published a container image for it. I'll re-publish one later today. I haven't had a chance yet to review the code, but it's possible your issue has already been addressed.

**Comment by sithmein** [comment]
OK, looking forward to the new images.
However, I believe it's the problem with missing properties discussed above and it's apparently more "severe". Here's what I did:

- Running `trivy` on the image directly returns the expected findings for `libc6` or `libc-bin`.
- Then I let `trivy` create an SBOM for this image and immediately ran it again on this SBOM - findings occur as well.
- Then I uploaded the SBOM to DT. No findings for `libc*` get reported when running through trivy (many others do)
- I exported the SBOM from DT and ran `trivy` on it - also here _no_ findings for `libc*`.

You can try it yourself using `registry.hub.knime.com/knime/knime-full:r-5.1.2-433` as the image.

**Comment by nscuro** [comment]
Updated images for API server and frontend have been pushed to Docker Hub.

> I believe it's the problem with missing properties discussed above and it's apparently more "severe".

This shortcoming is covered by #3369 already then.
